### PR TITLE
Update Director with provisioning Runtime information

### PR DESCRIPTION
**Pull request by franpog859**
**Description**

Changes proposed in this pull request:

- Update Director with Runtime Status Conditions which are `PROVISIONING`, `CONNECTED` and `FAILED`. `INITIAL` is default Status Condition
- Update Director with Runtime labels which are: `gardenerClusterDomain` and `gardenerClusterName`

**Related issue(s)**

See the issue #1149

### CODE REVIEW

File: components/provisioner/internal/operations/stages/wait_for_agent_to_connect_test.go [hunk 1]
```diff
@@ -54,6 +60,29 @@ func TestWaitForAgentToConnect(t *testing.T) {
...
+			// given
+			clientProvider := NewMockClientProvider(&v1alpha12.CompassConnection{
+				ObjectMeta: v1.ObjectMeta{Name: defaultCompassConnectionName},
+				Status: v1alpha12.CompassConnectionStatus{
+					State: testCase.state,
+				},
+			})
+
+			directorClient := &directorMocks.DirectorClient{}
+			directorClient.On("SetRuntimeStatusCondition", mock.AnythingOfType("string"), gqlschema.RuntimeStatusConditionConnected, mock.AnythingOfType("string")).Return(fmt.Errorf("some error"))
```
[the remark this answers is not in the document]
Comment (franpog859) [reply components/provisioner/internal/operations/stages/wait_for_agent_to_connect_test.go]: Done!

File: components/provisioner/internal/operations/stages/wait_for_agent_to_connect.go [hunk 2]
```diff
@@ -86,16 +94,29 @@ func (s *WaitForAgentToConnectStep) Run(cluster model.Cluster, _ model.Operation
 	if compassConnCR.Status.State != v1alpha1.Synchronized {
 		if compassConnCR.Status.State == v1alpha1.SynchronizationFailed {
 			logger.Warnf("Runtime Agent Connected but resource synchronization failed state: %s", compassConnCR.Status.State)
+			if err := s.directorClient.SetRuntimeStatusCondition(cluster.ID, gqlschema.RuntimeStatusConditionConnected, cluster.Tenant); err != nil {
+				logger.Errorf("Failed to set runtime %s status condition: %s", gqlschema.RuntimeStatusConditionConnected.String(), err.Error())
+				return operations.StageResult{Stage: s.Name(), Delay: 2 * time.Second}, nil
+			}
```
[the remark this answers is not in the document]
Comment (franpog859) [reply components/provisioner/internal/operations/stages/wait_for_agent_to_connect.go]: Done!

File: components/provisioner/pkg/gqlschema/schema.graphql [hunk 3]
```diff
@@ -125,6 +132,7 @@ input RuntimeInput {
     name: String!           # Name of the Runtime
     description: String     # Runtime description
     labels: Labels
+    statusCondition: RuntimeStatusCondition
```
[the remark this answers is not in the document]
Comment (franpog859) [reply components/provisioner/pkg/gqlschema/schema.graphql]: True ^^

File: components/provisioner/internal/graphql/gql_client_testkit.go [hunk 4]
```diff
@@ -27,16 +27,19 @@ func (c *QueryAssertClient) Do(req *graphql.Request, res interface{}) error {
 		c.expectedRequests = c.expectedRequests[1:]
 	}
 
-	if !c.shouldFail {
-		c.modifyResponseFunc(c.t, res)
+	if !c.shouldFail && len(c.modifyResponseFunc) > 0 {
+		c.modifyResponseFunc[0](c.t, res)
+		if len(c.modifyResponseFunc) > 1 {
+			c.modifyResponseFunc = c.modifyResponseFunc[1:]
+		}
```
[the remark this answers is not in the document]
Comment (franpog859) [reply components/provisioner/internal/graphql/gql_client_testkit.go]: Done!

File: components/provisioner/internal/gardener/step_provisioning_initial.go [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+	if err != nil {
+		log.Errorf("Error preparing Runtime Input: %s", err.Error())
+		return ctrl.Result{}, err
+	}
+	if err := r.directorClient.UpdateRuntime(runtimeId, runtimeInput, tenant); err != nil {
+		log.Errorf("Error updating Runtime in Director: %s", err.Error())
+		return ctrl.Result{}, err
+	}
+
+	log.Infof("Updating Shoot...")
```
[the remark this answers is not in the document]
Comment (franpog859) [reply components/provisioner/internal/gardener/step_provisioning_initial.go]: Done!

File: components/provisioner/internal/gardener/step_provisioning_initial.go [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+	log *logrus.Entry,
+	shoot gardener_types.Shoot,
+	operationId, runtimeId string) (ctrl.Result, error) {
+
+	if shoot.Spec.DNS == nil || shoot.Spec.DNS.Domain == nil {
+		log.Warnf("DNS Domain is not set yet for runtime ID: %s", runtimeId)
+		return ctrl.Result{}, nil
+	}
+
+	log.Infof("Updating in Director...")
```
[the remark this answers is not in the document]
Comment (franpog859) [reply components/provisioner/internal/gardener/step_provisioning_initial.go]: Done!

File: components/provisioner/internal/operations/executor.go [hunk 7]
```diff
@@ -70,6 +76,11 @@ func (e *Executor) Execute(operationID string) ProcessingResult {
 				log.Errorf("unrecoverable error occurred while processing operation: %s", err.Error())
 				e.handleOperationFailure(operation, cluster, log)
 				e.updateOperationStatus(log, operation.ID, nonRecoverable.Error(), model.Failed, time.Now())
+
+				if err := e.directorClient.SetRuntimeStatusCondition(cluster.ID, gqlschema.RuntimeStatusConditionFailed, cluster.Tenant); err != nil {
+					log.Errorf("failed to set runtime %s status condition: %s", gqlschema.RuntimeStatusConditionFailed.String(), err.Error())
+					return ProcessingResult{Requeue: true, Delay: defaultDelay}
+				}
```
[the remark this answers is not in the document]
Comment (franpog859) [reply components/provisioner/internal/operations/executor.go]: Done!

File: components/provisioner/pkg/gqlschema/schema.graphql [hunk 8]
```diff
@@ -111,6 +111,13 @@ enum OperationState {
     Failed
 }
 
+enum RuntimeStatusCondition {
+    INITIAL
+    PROVISIONING
+    CONNECTED
+    FAILED
+}
```
Comment (Szymongib) [inline components/provisioner/pkg/gqlschema/schema.graphql:119]: I think this can be removed also ;)

Comment (franpog859) [reply components/provisioner/pkg/gqlschema/schema.graphql]: True! Removed

**Pull request merged**